🚀 3단계 - 관리자 기능

hhhhhwi opened the pull request:
안녕하세요 😄 
3단계 PR 올립니다!

MemberRole Enum 클래스를 통해 권한에 대해 정의하고,
해당 값을 Member와 LoginMember에서 인스턴스 변수로 갖고 관리하고
MemberRoleInterceptor가 token의 role 값을 조회해 판단하도록 구현해보았습니다!

권한 체크 시, MemberRole 변환에 대한 exception 테스트를 하기 위해 예외 member 케이스를 db에 저장해두었어야 했는데요..!
그래서 일단 @Sql 으로 test.sql 을 실행하여 해당 테스트 클래스 실행 시 테스트 데이터를 생성시켰습니다!

추가적으로 피드백 남겨주시면 이번에도 열심히 🦾 ! 
적용해보도록 하겠습니다~!

hhhhhwi committed:
feat: Member의 권한에 대한 정보를 담은 MemberRole 생성
- Member 테이블에 컬럼 추가
- Member, LoginMember 인스턴스 변수로 MemberRole 추가

hhhhhwi committed:
feat: 토큰 발생 시, 권한 정보를 포함
- 잘못된 권한 정보 값에 대한 Custom Exception 생성

feat:

hhhhhwi committed:
test: 권한 체크에 대한 테스트 코드 작성

hhhhhwi committed:
test: /admin/** 경로에 대한 Interceptor 생성
- MemberRole.ADMIN의 경우에만 진입할 수 있도록 권한 체크

hhhhhwi committed:
fix: 응답코드 체크 버그 수정

hhhhhwi committed:
refactor: 코드 값 401 HttpStatus로 치환

boorownie commented on src/test/java/roomescape/test/LoginAcceptanceTest.java:
```diff
@@ -12,12 +12,14 @@
 @DisplayName("로그인 관련 api 호출 테스트")
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
 @DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+@Sql("/test.sql")
```
테스트를 위한 초기값 설정 방법은 다양한데요,
저도 회원과 역할과 같은 정보는 디비에 바로 넣는걸 선호합니다!

하지만 클래스나 비즈니스 로직의 수정 시 함께 챙겨줘야한다는 점이 있으니 참고해주세요!

boorownie commented on src/main/resources/schema.sql:
```diff
@@ -37,4 +38,5 @@ CREATE TABLE reservation
 
 
-INSERT INTO member (email, password, name) VALUES ('test@test.com', 'password', 'test');
+INSERT INTO member (email, password, name, role) VALUES ('test@test.com', 'password', 'test', 'member');
+INSERT INTO member (email, password, name, role) VALUES ('admin@test.com', 'password', 'admin', 'admin');
```
디비에 값을 바로 넣는 방법도 있지만
repository를 이용해서 save하는 방법, service 클래스를 이용하여 저장하는 방법 등 로직을 이용하는 방법도 있어요!
어떤 차이가 있을까요?

hhhhhwi replied:
스크립트를 사용하여 db에 등록하면 별 다른 설정 없이 바로 쿼리로 등록할 수 있어 편리하지만
지금처럼 member에 role 과 같은 새로운 컬럼이 추가됐을 때 인지하지 못 하면 컴파일 에러가 날 수 있다는 지점이 문제가 될 수 있다고 생각합니다! (ㅎㅎ 실제로 이번에 구현하면서 그랬습니다 😓)

그래서 ApplicationRunner로 애플리케이션 시작 시점에 데이터 등록하도록 수정해보았습니다!

boorownie commented on src/main/java/roomescape/login/MemberRoleInterceptor.java:
```diff
@@ -0,0 +1,44 @@
+            .orElseThrow(AuthenticationException::new)
+            .getValue();
+
+        MemberRole role = jwtTokenProvider.getMemberRole(token);
+
+        if(!role.equals(MemberRole.ADMIN)) {
+            response.setStatus(HttpStatus.UNAUTHORIZED.value());
+            return false;
+        }
```
인터셉터를 이용하여 어드민 권한 제한 👍 
토큰에 권한 정보를 넣고 권한 정보만 확인하는 로직 구성 좋네요!

boorownie commented on src/test/java/roomescape/step/LoginStep.java:
```diff
@@ -5,12 +5,21 @@
             .cookie("token");
     }
+
+    public static String 관리자_토큰_생성() {
```
현재는 로그인이라는 메서드를 이용해서 쿠키를 조회하는 방식으로 재사용을 하고 계신데요
만약 사용자가 여러명일 땐 아래 부분을 재사용할 수 있도록 메서드 분리를 할 수 있어요!
권한별 요청을 하는데 수월해질 수 있습니다! 참고해주세요~

```
RestAssured.given().log().all()
            .contentType(ContentType.JSON)
            .body(new LoginRequest("admin@test.com", "password"))
```

hhhhhwi replied:
중복 코드를 분리하여 해당 메소드를 같이 사용하도록 리팩토링 해보았습니다!

boorownie commented on src/test/java/roomescape/test/ThemeAcceptanceTest.java:
```diff
@@ -12,7 +12,6 @@
-import static roomescape.step.LoginStep.토큰_생성;
 import static roomescape.step.ReservationStep.예약_등록;
 
 @DisplayName("테마 관련 api 호출 테스트")
```
테마 관련 API 중 어드민만 요청해야 하는 API도 있을 수 있어요!
이 경우 테스트를 구축할 때 조금 고려해야할 부분이 있는데요,

어드민이 호출하는 상황과 일반 사용자가 호출하는 상황 두 경우를 테스트 해보면서
테스트 작성 및 코드 리팩터링을 경험해볼 수 있습니다!

boorownie reviewed (changes requested):
안녕하세요!
3단계 미션 구현 잘 해주셨어요!
사실 이미 충분히 잘 구현해주셔서 넘어가도 좋지만
테스트 코드 관련해서 하나 더 시도해보면 좋을 것 같아서 코멘트 남겨봅니다!

혹시나 미션3에서 시도해보고 싶으시면 DM 으로 말씀해주시고요~
가볍게 시도해볼 수 있으실 것 같으면 다시 리뷰요청 해주시면 감사하겠습니다!

hhhhhwi committed:
refactor: 중복 메서드 분리

hhhhhwi committed:
refactor: CookieUtils 클래스 생성으로 쿠키 생성 중복 코드 제거

hhhhhwi committed:
test: 테마 등록, 삭제 api는 관리자 권한만 실행 가능함 에 대한 test 메소드 수정

hhhhhwi committed:
refactor: MemberRoleInterceptor의 pattern 추가
- /themes/**, /times/**

hhhhhwi committed:
fix: /themes/**, /times/** URI 패턴 중 권한 체크 불필요한 api에 예외처리
- 테마 조회 : GET /themes
- 시간 조회 : GET /times
- 예약 가능 시간 조회 : GET /times/available

hhhhhwi committed:
refactor: 예약 테마, 예약 시간 등록 재사용을 위한 static 메소드 생성
- ThemeStep, ReservationTimeStep
- 예약 시간 등록, 삭제에 대한 권한 체크 테스트 메소드 추가

hhhhhwi committed:
refactor: 더미 데이터 등록 방법 변경
- ApplicationRunner를 활용하여 회원 데이터 등록
- 테스트 코드에서도 해당 데이터를 활용할 수 있도록 회원 정보 상수화

hhhhhwi commented on src/main/java/roomescape/login/MemberRoleInterceptor.java:
```diff
@@ -4,14 +4,16 @@
 
 public class MemberRoleInterceptor implements HandlerInterceptor {
 
+    private final List<String> excludeGetUrls = List.of("/themes", "/times", "/times/available");
```
> 테마 관련 API 중 어드민만 요청해야 하는 API도 있을 수 있어요!

해당 api에 대해 interceptor 패턴을 등록을 하니, 
조회 메소드는 사용자 페이지와 같이 쓰고 있다는 점에서 문제가 생겼습니다

애초에 위와 같은 패턴이 안 나오도록 경로를 정하는 것이 중요하지만, 막상 운영하다보면 애매한 상황이 늘 생기더라구요..😞 

uri 경로 변경 및 관리자와 사용자 용 메소드 분리 처리 등 확실한 방법이 있지만
우선 수정 범위를 적은 방법인 해당 uri만 예외처리 하는 방식으로 구현해보았습니다..!

boorownie replied:
네 맞습니다. 경로 변경이 확실한 방법입니다만 수정 범위를 줄이는 것도 접근이라고 생각합니다 👍

hhhhhwi commented on src/main/java/roomescape/member/initializer/MemberInitializer.java:
```diff
@@ -0,0 +1,31 @@
+import roomescape.member.MemberRole;
+import roomescape.member.service.MemberRepository;
+
+@Component
+public class MemberInitializer implements ApplicationRunner {
```
스크립트를 사용하여 db에 등록하면 별 다른 설정 없이 바로 쿼리로 등록할 수 있어 편리하지만
지금처럼 member에 role 과 같은 새로운 컬럼이 추가됐을 때 인지하지 못 하면 컴파일 에러가 날 수 있다는 지점이 문제가 될 수 있다고 생각합니다! (ㅎㅎ 실제로 이번에 구현하면서 그랬습니다 😓)

그래서 ApplicationRunner로 애플리케이션 시작 시점에 데이터 등록하도록 수정해보았습니다!

boorownie replied:
현재 MemberInitializer에는 테스트를 위한 부분도 포함되어있는데
테스트를 위한 초기화와 프로덕션을 위한 초기화를 구분하는 것도 방법일 것 같습니다!
이 경우 Profile을 활용하는 것도 방법이 될 수 있겠네요 :)

boorownie commented on src/test/java/roomescape/step/LoginStep.java:
```diff
@@ -5,19 +5,24 @@
-            .when().post("/login")
-            .then().log().all().extract()
-            .cookie("token");
+        return 토큰_생성(회원_이메일, 회원_비밀번호);
```
피드백 반영 👍

boorownie reviewed (approved):
피드백 반영 잘 해주셨습니다!
고민해보실 수 있는 부분에 남긴 코멘트 모두 잘 고민해주셨네요 👍 

이번 미션은 이만 머지하겠습니다!
남은 미션3도 끝까지 완주하시길 바라봅니다 🙏
